File: client/MemDB-Java-Client/src/test/java/es/memdb/messages/response/ResponseDeserializerTest.java
package es.memdb.messages.response;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

public class ResponseDeserializerTest  {
    private ResponseDeserializer responseDeserializer;

    @Before
    public void setup() {
        this.responseDeserializer = new ResponseDeserializer();
    }

    @Test
    public void shouldDeserializeError() {
        byte[] toDeserialize = new byte[]{
                0x00, 0x00, 0x00, 0x0D, //Response total length
                0x00, 0x00, 0x00, 0x01, //Req num
                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, //Timestamp
                0x02 //Error
        };

        Response response = this.responseDeserializer.deserialize(toDeserialize);

        Assert.assertFalse(response.isSuccessful());
        Assert.assertEquals(1, response.getTimestamp());
        Assert.assertEquals(1, response.getRequestNumber());
        Assert.assertEquals(1, response.getErrorCode());
        Assert.assertEquals("", response.getResponse());
    }

    @Test
    public void shouldDeserializeSuccess() {
        byte[] toDeserialize = new byte[]{
                0x00, 0x00, 0x00, 0x13, //Response total length
                0x00, 0x00, 0x00, 0x01, //Req num
                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, //Timestamp
                0x01, //Success
                0x00, 0x00, 0x00, 0x02, //Response body length
                0x4F, 0x4B //Response body
        };

        Response response = this.responseDeserializer.deserialize(toDeserialize);

        Assert.assertTrue(response.isSuccessful());
        Assert.assertEquals(1, response.getTimestamp());
        Assert.assertEquals(1, response.getRequestNumber());
        Assert.assertEquals("OK", response.getResponse());
    }
}

File: core/tst/messages/response/ResponseDeserializerTest.cpp
#include "gtest/gtest.h"

#include "messages/response/ResponseDeserializer.h"
#include "../../Assertions.h"

#include <string>

TEST(ResponseDeserializer, ResponseWithArgSuccess) {
    ResponseDeserializer responseDeserializer{};
    auto responseDeserialized = responseDeserializer.deserialize({
        0x00, 0x00, 0x00, 0x02, //Request number
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, //Timestamp
        0x01, //Success
        0x00, 0x00, 0x00, 0x05, //Length response
        0x68, 0x65, 0x6C, 0x6C, 0x6F //hello
    });

    ASSERT_EQ(responseDeserialized.requestNumber, 2);
    ASSERT_EQ(responseDeserialized.timestamp, 1);
    ASSERT_TRUE(responseDeserialized.isSuccessful);

    ASSERT_EQ(responseDeserialized.responseValue.size, 5);

    std::vector<uint8_t> expectedArgValue = {0x68, 0x65, 0x6C, 0x6C, 0x6F};
    Assertions::assertPointerValues(expectedArgValue.data(), responseDeserialized.responseValue.data(),
                                    responseDeserialized.responseValue.size);
}

TEST(ResponseDeserializer, ResponseWithoutArgsError) {
    ResponseDeserializer responseDeserializer{};
    auto responseDeserialized = responseDeserializer.deserialize({
        0x00, 0x00, 0x00, 0x02, //Request number
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, //Timestamp
        0x02, //_error errorcode 1
        0x00, 0x00, 0x00, 0x00, //Length response
    });

    ASSERT_EQ(responseDeserialized.requestNumber, 2);
    ASSERT_EQ(responseDeserialized.timestamp, 1);
    ASSERT_FALSE(responseDeserialized.isSuccessful);
    ASSERT_EQ(responseDeserialized.errorCode, 1);

    ASSERT_EQ(responseDeserialized.responseValue.size, 0);
}